# Introduction

We have to learn some basic programming concepts before doing anything in Maching Learning or Data Science or any Projects 

# Python Basics

## Entering expressions

In [1]:
2+2

4

**Commonly used math operators**
- Addition 
- Subtraction
- Multiplication
- Division
- Integer Division
- Moduls/ Remainder
- Exponent/ Power

**Order of Precedence**

The order of operations (also called precedence) of Python math operators is similar to that of mathematics. The ** operator is evaluated first; the *, /, //, and % operators are evaluated next, from left to right; and the + and - operators are evaluated last (also from left to right)

In [15]:
2+3*6

20

In [16]:
(2 + 3) * 6

30

In [17]:
2**5

32

In [25]:
22/7

3.142857142857143

In [26]:
22//7

3

In [27]:
22 % 7

1

## Common data types

**Common Data Types**
- Integer - `int`
    - e.g. `0,50,-100`
- Floating-Point - `float`
    - e.g. `1.25, 3.14, -1.0, --0.5, 0.0`
- String - `str`
    - e.g. `'a', 'aa', 'aaa', 'Hello!', '11 cats'`
- Boolean - `bool`
    - e.g. `True, False`

String can be enclosed in single or double quotes

## String concatenation & Replicaiton

The meaning of an operator (e.g. `+`) may change based on the data types of the values next to it


In [23]:
# Addition for int or float
1 + 2

3

In [22]:
# Concatenation for string data type
'Bruce' + 'Wayne'

'BruceWayne'

if you try to use the + operator on a string and an integer value, Python will not know how to handle this, and it will display an <font color = red>error message</font>.

In [29]:
'Bruce' + 42

TypeError: must be str, not int

The `*` operator when used on one string value and one integer value becomes string replication operator.

In [31]:
2*5

10

In [30]:
'Batman'*5

'BatmanBatmanBatmanBatmanBatman'

**Varaibles**
- It's like box in the computer’s memory where we can store values

**Variable Names** are case sensitive.<br>We  can name a variable anything as long as it obeys the following three rules:

- It can be only one word.
- It can use only letters, numbers, and the underscore `_` character.
- It can’t begin with a number.

e.g. of *invalid variable names*: 	
`current-balance, current balance 4account, total_$um, 'batman'`

## `Hello world` - First Python program

In [4]:
# This program says hello and asks for my name.

print('Hello world!')
print('What is your name?')    # ask for their name
myName = input()
print('It is good to meet you, ' + myName)
print('The length of your name is:')
print(len(myName))
print('What is your age?')    # ask for their age
myAge = input()
print (type(myAge))
print('You will be ' + str(int(myAge) + 1) + ' in a year.')

Hello world!
What is your name?
James Bond
It is good to meet you, James Bond
The length of your name is:
10
What is your age?
50
<class 'str'>
You will be 51 in a year.


### Program Breakdown

The following line is called a comment. Python ignores comments, and you can use them to write notes or remind yourself what the code is trying to do. Any text for the rest of the line following a hash mark `#` is part of a comment.

#### Comments

In [5]:
# This program says hello and asks for my name.

The `print()` function displays the `string` value inside the parentheses on the screen. Notice that the quotes are not printed to the screen. They just mark where the string begins and ends; they are not part of the string value.

#### `print()` function

In [6]:
print('Hello world!')
print('What is your name?')    # ask for their name

Hello world!
What is your name?


The `input()` function waits for the user to type some text on the keyboard and press ENTER key.
The user input will then be stored in `myName` variable

#### `input()` function

In [7]:
myName = input()

Gandolf the Gray


The expression `'It is good to meet you, ' + myName` is evaluated and is `print`ed on the screen.

In [8]:
print('It is good to meet you, ' + myName)

It is good to meet you, Gandolf the Gray


The `len()` function returns the number of characters in the given string.

#### `len()` function

In [9]:
print('The length of your name is:')
print(len(myName))

The length of your name is:
16


#### type() function

The `type()` function returns the data type of the value given

In [7]:
print('What is your age?')    # ask for their age
myAge = input()
print (type(myAge))           # the input functions usually returns a string data type

What is your age?
50
<class 'str'>


#### `str(), int(), float()` functions

These functions are used to change the data types of the values. <br>
As the `input()` function returns `str` data, `myAge` variable is of `str` data type. <br>
To add a numerical value to it, the value should be converted to `int`.<br> 
After performing the addition, the `int` value is converted back to `str` again so that it can be concatenated with other strings.

In [8]:
print('You will be ' + str(int(myAge) + 1) + ' in a year.')

50


# Flow control

## `and,or,not` and `<,>,<=,>=,==,!=`

- Boolean Values: `True, False`
- Boolean operators: `and,or,not`
    - returns `True` or `False`
- Comparison operators: `<, >, <=, >=, ==, !=`
    - returns `True` or `False`

In [14]:
5 > 6

(False, True, False)

In [20]:
5 <= 6, 3>5, 3!=5, 3 == 5

(True, False, True, False)

In [17]:
5<6 and 3>5, 5<6 or 3>5

(False, True)

In [12]:
not (True), not(False)

(False, True)

## Flow control

### if

- When `if` condition evalues to `True`, the block of statement indented inside will get executed

In [30]:
age = 5
if age < 10:                       # age < 10 evaluates to True      
    print ('age is less than 10')

age is less than 10


In [31]:
age = 15
if age < 10:                       # age < 10 evaluates to False             
    print ('age is less than 10')  # Doesn't get printed as age variable us more than 10

### if else

- When `if` condition evalues to `True`, the block of statement indented inside `if` will get executed.
- When it evaluates to `False`, the  block of statement indented inside `else` will get executed.

In [32]:
age = 18
if age < 18:                     
    print ('You are not an adult')
else:
    print ('You are an adult')

You are an adult


### `if elif else`

- When `if` condition evalues to `True`, the block of statement indented inside `if` will get executed.
- When it fails, the `elif` condition will be evaluated and if it turns out to be `True`,  the block of statement indented inside `elif` will get executed.
- When both `if` and `elif` evaluates to `False`, the  block of statement indented inside `else` will get executed.

In [37]:
age = 10
if age > 18:
    print ('You are  an adult')
elif age > 12:
    print ('You are a teenager')
else:
    print ('You are neither adult nor teen. Just a kid!')

You are neither adult nor teen. Just a kid!


We can have multiple `elif` statements after an `if` statement. In the entire combo, `if-elseif-else` combo, only one of the condition will be exeucted.

In [41]:
age = 15
if age >60 :
    print ('You are very old')
elif age>18:
    print ('You are an adult')
elif age > 12:
    print ('You are a teenager')
else:
    print ('You are neither adult nor teen. Just a kid!')

You are a teenager


### nested `if`

We can have `if-else` inside another `if-else` and we can keep nesting them. The deeper the nest gets, the harder it will be to debug and trace.

In [44]:
user_name = input('Enter user name: ')      # james bond
pass_word = input('Enter pass word: ')      # 007
if user_name == 'james bond':
    if pass_word == '007':
        print ('Access granted')
    else:
        print ('Access denied. Wrong password')
else:
     print ('Access denied. Wrong username')
        

Enter user name: james bond
Enter pass word: 008
Access denied. Wrong password


The above code could be re-witten as follows using `boolean` and `comparison operators` making it easy to read and follow

In [46]:
user_name = input('Enter user name: ')      # james bond
pass_word = input('Enter pass word: ')      # 007
if user_name == 'james bond' and pass_word == '007':
    print ('Access granted')
else:
    print ('Access denied. Wrong login details')

Enter user name: james bond
Enter pass word: 007
Access granted
